subprojects/model-core/src/integTest/groovy/org/gradle/api/provider/PropertyAssignmentIntegrationTest.groovy

```diff
@@ -0,0 +1,405 @@
+        "File = T extends FileSystemLocation"           | "File"         | "DirectoryProperty" | 'layout.buildDirectory.dir("out").get()' | failsWithDescription("Type mismatch") | "Result: out"
+        "File = Provider<T extends FileSystemLocation>" | "File"         | "DirectoryProperty" | 'layout.buildDirectory.dir("out")'       | failsWithDescription("Type mismatch") | "Result: out"
+        "File = File"                                   | "File"         | "DirectoryProperty" | 'file("$buildDir/out")'                  | "Result: out"                         | "Result: out"
+        "File = Provider<File>"                         | "File"         | "DirectoryProperty" | 'provider { file("$buildDir/out") }'     | failsWithDescription("Type mismatch") | "Result: out"
+        "File = Object"                                 | "File"         | "DirectoryProperty" | 'MyObject("out")'                        | failsWithDescription("Type mismatch") | failsWithDescription("Type mismatch")
+    }
+
+    def "test Groovy collection types assignment for #description"() {
+        given:
+        def buildFileDefinition = """
+            abstract class MyTask extends DefaultTask {
+                @Internal
+                {inputDeclaration}
+
+                @TaskAction
+                void run() {
+                    if (input instanceof Provider) {
+                        println("Result: " + input.get())
+                    } else {
+                        println("Result: " + input)
+                    }
+                }
+            }
+
+            tasks.register("myTask", MyTask) {
+                input $operation {inputValue}
+            }
+        """
+
+        when:
+        def defaultEagerInputValue = eagerInputType.contains("Map<") ? "[:]" : "[]"
+        buildFile.text = buildFileDefinition
+            .replace("{inputDeclaration}", "$eagerInputType input = $defaultEagerInputValue")
+            .replace("{inputValue}", inputValue)
+
+        then:
+        runAndAssert("myTask", eagerResult)
+
+        when:
+        buildFile.text = buildFileDefinition
+            .replace("{inputDeclaration}", "abstract $lazyInputType getInput()")
+            .replace("{inputValue}", inputValue)
+
+        then:
+        runAndAssert("myTask", lazyResult)
+
+        where:
+        description                              | operation | eagerInputType        | lazyInputType                 | inputValue                               | eagerResult                              | lazyResult
+        "Collection<T> = T[]"                    | "="       | "List<String>"        | "ListProperty<String>"        | '["a"] as String[]'                      | 'Result: [a]'                            | failsWithCause("Cannot set the value of a property of type java.util.List using an instance of type [Ljava.lang.String;")
```

Comment:
:x: Let's not use `String` here either, but `MyObject` or maybe `Integer` or some other type that does not get auto-conversion.